File: python/rest-api/rest_api.py
import json

class RestAPI:
    def __init__(self, database=None):
        # Initialize the RestAPI with an optional database.
        self.database = database

    def get(self, url, payload=None):
        # Parse the payload if it's provided, or initialize it with an empty "users" list.
        payload = json.loads(payload) if payload is not None else {"users": []}

        if url == "/users":
            # Filter users from the database based on the names provided in the payload.
            users = [user for user in self.database["users"]
                     if user["name"] in payload["users"]]

            # Return a JSON response containing the sorted list of users by name.
            return json.dumps({"users": sorted(users,
                                               key=lambda user: user["name"])})

    def post(self, url, payload=None):
        # Parse the payload if it's provided, or initialize it with an empty "users" list.
        payload = json.loads(payload) if payload is not None else {"users": []}

        if url == "/add":
            # Add a new user to the database with the given name and initialize their data.
            user = {"name": payload["user"],
                    "owes": {}, "owed_by": {}, "balance": 0.0}
            self.database["users"].append(user)

            # Return the user data as a JSON response.
            return json.dumps(user)
        elif url == "/iou":
            # Extract lender, borrower, and amount from the payload.
            lender = payload["lender"]
            borrower = payload["borrower"]
            amount = payload["amount"]
            users = {"users": []}

            for user in self.database["users"]:
                if user["name"] == lender:
                    # Update the lender's balance and handle debts/credits.
                    user["balance"] += amount
                    remain = amount
                    users["users"].append(user)

                    if borrower in user["owes"]:
                        if user["owes"][borrower] > remain:
                            user["owes"][borrower] -= remain
                            remain = 0
                        elif user["owes"][borrower] <= remain:
                            remain -= user["owes"][borrower]
                            user["owes"].pop(borrower)

                    if remain > 0:
                        if borrower not in user["owed_by"]:
                            user["owed_by"][borrower] = 0
                        user["owed_by"][borrower] += remain
                elif user["name"] == borrower:
                    # Update the borrower's balance and handle debts/credits.
                    user["balance"] -= amount
                    remain = amount
                    users["users"].append(user)

                    if lender in user["owed_by"]:
                        if user["owed_by"][lender] > remain:
                            user["owed_by"][lender] -= remain
                            remain = 0
                        elif user["owed_by"][lender] <= remain:
                            remain -= user["owed_by"][lender]
                            user["owed_by"].pop(lender)

                    if remain > 0:
                        if lender not in user["owes"]:
                            user["owes"][lender] = 0
                        user["owes"][lender] += remain

            # Return the updated user data as a JSON response.
            return json.dumps(users)

File: javascript/rest-api/rest-api.js
export class RestAPI {
  constructor(database = {}) {
    this.database = database;
  }

  get(url) {
    const [endpoint, queryParams] = url.split("?");
    const queryObj = this.parseQueryParams(queryParams);

    switch (endpoint) {
      case "/users":
        return this.getUsers(queryObj.users);
      default:
        return { error: "Endpoint not found" };
    }
  }

  post(url, payload) {
    const [endpoint] = url.split("/");

    switch (endpoint) {
      case "add":
        return this.createUser(payload.user);
      case "iou":
        return this.createIOU(payload.lender, payload.borrower, payload.amount);
      default:
        return { error: "Endpoint not found" };
    }
  }

  parseQueryParams(queryParams) {
    if (!queryParams) return {};
    const params = new URLSearchParams(queryParams);
    const result = {};
    for (const [key, value] of params) {
      result[key] = value.split(",");
    }
    return result;
  }

  getUsers(users) {
    const userList = Object.keys(this.database).sort();
    const filteredUsers = userList.filter(
      (user) => !user || user.includes(user),
    );
    const userObjects = filteredUsers.map((user) => this.getUserObject(user));
    return { users: userObjects };
  }

  createUser(userName) {
    if (this.database[userName]) {
      return { error: "User already exists" };
    }
    this.database[userName] = {
      name: userName,
      owes: {},
      owed_by: {},
      balance: 0.0,
    };
    return this.getUserObject(userName);
  }

  createIOU(lender, borrower, amount) {
    if (!this.database[lender] || !this.database[borrower]) {
      return { error: "Invalid lender or borrower" };
    }

    amount = parseFloat(amount);
    if (isNaN(amount) || amount <= 0) {
      return { error: "Invalid amount" };
    }

    this.database[lender].owes[borrower] =
      (this.database[lender].owes[borrower] || 0) + amount;
    this.database[borrower].owed_by[lender] =
      (this.database[borrower].owed_by[lender] || 0) + amount;
    this.updateBalances(lender);
    this.updateBalances(borrower);

    return {
      users: [this.getUserObject(lender), this.getUserObject(borrower)],
    };
  }

  getUserObject(userName) {
    const user = this.database[userName];
    const balance = this.calculateBalance(userName);
    return {
      name: user.name,
      owes: user.owes,
      owed_by: user.owed_by,
      balance: balance,
    };
  }

  calculateBalance(userName) {
    const user = this.database[userName];
    const owesTotal = Object.values(user.owes).reduce(
      (total, amount) => total + amount,
      0,
    );
    const owedByTotal = Object.values(user.owed_by).reduce(
      (total, amount) => total + amount,
      0,
    );
    return owedByTotal - owesTotall;
  }

  updateBalances(userName) {
    this.database[userName].balance = this.calculateBalance(userName);
  }
}
